File: backends/node/src/email.ts
import nodemailer, { type Transporter } from 'nodemailer';

export interface MailerConfig {
  host: string;
  port: number;
  from: string;
  publicBaseUrl: string;
}

export class Mailer {
  private readonly transport: Transporter;
  private readonly from: string;
  private readonly publicBaseUrl: string;

  constructor(config: MailerConfig) {
    this.transport = nodemailer.createTransport({
      host: config.host,
      port: config.port,
      secure: false,
      ignoreTLS: true,
    });
    this.from = config.from;
    this.publicBaseUrl = config.publicBaseUrl;
  }

  shareUrl(token: string): string {
    return `${this.publicBaseUrl}/share/${token}`;
  }

  async sendShareLinkEmail(args: {
    to: string;
    orgName: string;
    ticketSubject: string;
    shareToken: string;
  }): Promise<void> {
    const url = this.shareUrl(args.shareToken);
    await this.transport.sendMail({
      from: this.from,
      to: args.to,
      subject: `Your support request — ${args.ticketSubject}`,
      text: [
        `Thanks for contacting ${args.orgName}.`,
        ``,
        `View your ticket and reply at:`,
        `  ${url}`,
        ``,
        `This link is valid for 30 days.`,
      ].join('\n'),
    });
  }

  async sendCustomerReplyNotification(args: {
    to: string[];
    orgName: string;
    ticketSubject: string;
    customerEmail: string;
    reopened: boolean;
  }): Promise<void> {
    if (args.to.length === 0) return;
    const subjectPrefix = args.reopened
      ? `Ticket reopened by customer reply`
      : `New customer reply`;
    await this.transport.sendMail({
      from: this.from,
      to: args.to.join(', '),
      subject: `${subjectPrefix} — ${args.ticketSubject}`,
      text: [
        `${args.customerEmail} replied to "${args.ticketSubject}" in ${args.orgName}.`,
        args.reopened ? `The ticket was resolved and is now reopened.` : ``,
        ``,
        `Open the ticket in your inbox to respond.`,
      ]
        .filter(Boolean)
        .join('\n'),
    });
  }
}

File: backends/python/src/hearth_python/email_.py
"""SMTP mailer for share-link and admin-notification emails."""

from __future__ import annotations

import smtplib
from dataclasses import dataclass
from email.message import EmailMessage


@dataclass(frozen=True)
class MailerConfig:
    host: str
    port: int
    from_addr: str
    public_base_url: str


class Mailer:
    def __init__(self, config: MailerConfig) -> None:
        self._config = config

    def share_url(self, token: str) -> str:
        return f"{self._config.public_base_url}/share/{token}"

    def send_share_link_email(
        self,
        *,
        to: str,
        org_name: str,
        ticket_subject: str,
        share_token: str,
    ) -> None:
        url = self.share_url(share_token)
        msg = EmailMessage()
        msg["From"] = self._config.from_addr
        msg["To"] = to
        msg["Subject"] = f"Your support request — {ticket_subject}"
        msg.set_content(
            "\n".join(
                [
                    f"Thanks for contacting {org_name}.",
                    "",
                    "View your ticket and reply at:",
                    f"  {url}",
                    "",
                    "This link is valid for 30 days.",
                ]
            )
        )
        self._send(msg)

    def send_customer_reply_notification(
        self,
        *,
        to: list[str],
        org_name: str,
        ticket_subject: str,
        customer_email: str,
        reopened: bool,
    ) -> None:
        if not to:
            return
        subject_prefix = "Ticket reopened by customer reply" if reopened else "New customer reply"
        msg = EmailMessage()
        msg["From"] = self._config.from_addr
        msg["To"] = ", ".join(to)
        msg["Subject"] = f"{subject_prefix} — {ticket_subject}"
        body_lines = [
            f'{customer_email} replied to "{ticket_subject}" in {org_name}.',
        ]
        if reopened:
            body_lines.append("The ticket was resolved and is now reopened.")
        body_lines.extend(["", "Open the ticket in your inbox to respond."])
        msg.set_content("\n".join(body_lines))
        self._send(msg)

    def _send(self, msg: EmailMessage) -> None:
        # local_hostname pinned to 'hearth-python' instead of letting
        # smtplib call socket.getfqdn() — on macOS that triggers a slow
        # reverse-DNS lookup that adds ~25s to every send.
        with smtplib.SMTP(
            self._config.host,
            self._config.port,
            local_hostname="hearth-python",
            timeout=10,
        ) as smtp:
            smtp.send_message(msg)
